rero_ils/modules/cli.py

```diff
@@ -475,3 +510,243 @@ def compile_json(src_jsonfile, output, verbose):
     if not output:
         output = sys.stdout
     json.dump(data, fp=output, indent=2)
+
+
+def do_worker(marc21records, results, pid_requierd):
+    """Worker for marc21 to json transformation."""
+    schema_in_bytes = resource_string(
+        'rero_ils.modules.documents.jsonschemas',
+        'documents/document-v0.0.1.json'
+    )
+    schema = loads(schema_in_bytes.decode('utf8'))
+    for data in marc21records:
+        data_json = data['json']
+        pid = data_json.get('001', '???')
+        try:
+            record = marc21tojson.do(data_json)
+            if not record.get("$schema"):
+                # create dummy schema in data
+                record["$schema"] = 'dummy'
+            if not pid_requierd:
+                if not record.get("pid"):
+                    # create dummy pid in data
+                    record["pid"] = 'dummy'
+            validate(record, schema)
+            if record["$schema"] == 'dummy':
+                del record["$schema"]
+            if not pid_requierd:
+                if record["pid"] == 'dummy':
+                    del record["pid"]
+            results.append({
+                'status': True,
+                'data': record
+            })
+        except Exception as err:
+            msg = ''
+            # make the error message shorter
+            for line in str(err).split('\n'):
+                if line and line[0] != ' ':
+                    if len(line) > 100:
+                        line = line[:25] + ' ...'
+                    msg += '\t' + line
+            msg = msg.strip()
+            print('ERROR:', pid, msg, sep='\t', file=sys.stderr)
+            # import traceback
+            # traceback.print_exc()
+            results.append({
+                'pid': pid,
+                'status': False,
+                'data': data['xml']
+            })
+
+
+class Marc21toJson():
+    """Class for Marc21 recorts to Json transformation."""
+
+    __slots__ = ['xml_file', 'json_file_ok', 'xml_file_error', 'parallel',
```

Comment:
this is for RAM optimization we can try to use pypy as it do it by default. See: http://book.pythontips.com/en/latest/__slots__magic.html
for more details

Reply:
Do we have to run the complete rero-ils app with pypy then?